signer/plugin.go

```diff
@@ -0,0 +1,260 @@
+		return s.generateSignatureEnvelope(ctx, desc, opts)
+	}
+	return nil, nil, fmt.Errorf("plugin does not have signing capabilities")
+}
+
+func (s *pluginSigner) generateSignature(ctx context.Context, desc ocispec.Descriptor, opts notation.SignOptions) ([]byte, *signature.SignerInfo, error) {
+	config := s.mergeConfig(opts.PluginConfig)
+	// Get key info.
+	key, err := s.describeKey(ctx, config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Check keyID is honored.
+	if s.keyID != key.KeyID {
+		return nil, nil, fmt.Errorf("keyID in describeKey response %q does not match request %q", key.KeyID, s.keyID)
+	}
+	ks, err := proto.DecodeKeySpec(key.KeySpec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if remoteSigner, ok := s.signer.(*remoteSigner); ok {
+		remoteSigner.ctx = ctx
+		remoteSigner.keySpec = ks
+		remoteSigner.pluginConfig = config
+	}
+
+	return generateSignatureBlob(ctx, s.signer, desc, opts)
+}
+
+func (s *pluginSigner) generateSignatureEnvelope(ctx context.Context, desc ocispec.Descriptor, opts notation.SignOptions) ([]byte, *signature.SignerInfo, error) {
+	payload := envelope.Payload{TargetArtifact: envelope.SanitizeTargetArtifact(&desc)}
+	payloadBytes, err := json.Marshal(payload)
+	if err != nil {
+		return nil, nil, fmt.Errorf("envelope payload can't be marshaled: %w", err)
+	}
+	// Execute plugin sign command.
+	req := &proto.GenerateEnvelopeRequest{
+		ContractVersion:       proto.ContractVersion,
+		KeyID:                 s.keyID,
+		Payload:               payloadBytes,
+		SignatureEnvelopeType: opts.SignatureMediaType,
+		PayloadType:           envelope.MediaTypePayloadV1,
+		PluginConfig:          s.mergeConfig(opts.PluginConfig),
+	}
+	resp, err := s.plugin.GenerateEnvelope(ctx, req)
+	if err != nil {
+		return nil, nil, fmt.Errorf("generate-envelope command failed: %w", err)
+	}
+
+	// Check signatureEnvelopeType is honored.
+	if resp.SignatureEnvelopeType != req.SignatureEnvelopeType {
+		return nil, nil, fmt.Errorf(
+			"signatureEnvelopeType in generateEnvelope response %q does not match request %q",
+			resp.SignatureEnvelopeType, req.SignatureEnvelopeType,
+		)
+	}
+
+	sigEnv, err := signature.ParseEnvelope(opts.SignatureMediaType, resp.SignatureEnvelope)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	envContent, err := sigEnv.Verify()
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := envelope.ValidatePayloadContentType(&envContent.Payload); err != nil {
+		return nil, nil, err
+	}
+
+	var signedPayload envelope.Payload
+	if err = json.Unmarshal(envContent.Payload.Content, &signedPayload); err != nil {
+		return nil, nil, fmt.Errorf("signed envelope payload can't be unmarshaled: %w", err)
+	}
+
+	// TODO: Verify plugin did not add any additional top level payload attributes. https://github.com/notaryproject/notation-go/issues/80
+	if !descriptorPartialEqual(desc, signedPayload.TargetArtifact) {
+		return nil, nil, errors.New("descriptor subject has changed")
```

Comment:
```suggestion
		return nil, nil, fmt.Errorf("during signing descriptor subject has changed from %s to %s", desc, signedPayload.TargetArtifact)
```
ps: you might need to serialize the descriptors

Reply:
> 

How about `return nil, nil, fmt.Errorf("during signing descriptor subject has changed from %+v to %+v", desc, signedPayload.TargetArtifact)`